onnxruntime/core/providers/shared_library/provider_interfaces.h

```diff
@@ -0,0 +1,480 @@
+  virtual ::onnx::Provider_TensorProto* add_tensors() = 0;
+
+  void operator=(const Provider_AttributeProto& v) = delete;
+};
+
+// This is needed since Provider_NodeAttributes is a map of unique_ptr to Provider_AttributeProto and that won't work since unique_ptrs are not copyable
+// (supposedly this should work in the latest C++ STL but it didn't for me so I used this to make it copyable)
+struct Provider_AttributeProto_Copyable {
+  Provider_AttributeProto_Copyable() = default;
+  Provider_AttributeProto_Copyable(const Provider_AttributeProto_Copyable& copy) : p_{copy->Clone()} {}
+
+  void operator=(std::unique_ptr<Provider_AttributeProto>&& p) { p_ = std::move(p); }
+  void operator=(const Provider_AttributeProto_Copyable& p) { p_ = p->Clone(); }
+
+  Provider_AttributeProto& operator*() const { return *p_.get(); }
+  Provider_AttributeProto* operator->() const { return p_.get(); }
+
+  std::unique_ptr<Provider_AttributeProto> p_;
+};
+
+struct Provider_TensorShapeProto {
+  int dim_size() const { return dim_size_; }
+
+  int dim_size_{};
+};
+
+}  // namespace ONNX_NAMESPACE
+
+namespace onnxruntime {
+
+struct ProviderHost;
+struct Provider_IExecutionProvider;
+
+struct Provider_IExecutionProviderFactory {
+  virtual ~Provider_IExecutionProviderFactory() = default;
+  virtual std::unique_ptr<Provider_IExecutionProvider> CreateProvider() = 0;
+};
+
+//struct KernelCreateInfo;
+
+class DataTypeImpl;
+using MLDataType = const DataTypeImpl*;
+
+struct Provider_OrtDevice {
+  virtual ~Provider_OrtDevice() {}
+};
+
+struct Provider_OrtMemoryInfo {
+  static std::unique_ptr<Provider_OrtMemoryInfo> Create(const char* name_, OrtAllocatorType type_, Provider_OrtDevice* device_ = nullptr, int id_ = 0, OrtMemType mem_type_ = OrtMemTypeDefault);
+  virtual ~Provider_OrtMemoryInfo() {}
+
+  void operator=(const Provider_OrtMemoryInfo& v) = delete;
+};
+
+template <typename T>
+using Provider_IAllocatorUniquePtr = std::unique_ptr<T, std::function<void(T*)>>;
+
+struct Provider_IAllocator {
+  virtual ~Provider_IAllocator() {}
+
+  virtual void* Alloc(size_t size) = 0;
+  virtual void Free(void* p) = 0;
+
+  static bool CalcMemSizeForArray(size_t nmemb, size_t size, size_t* out) noexcept {
+    return CalcMemSizeForArrayWithAlignment<0>(nmemb, size, out);
+  }
+
+  template <size_t alignment>
+  static bool CalcMemSizeForArrayWithAlignment(size_t nmemb, size_t size, size_t* out) noexcept ORT_MUST_USE_RESULT;
+
+  template <typename T>
+  static Provider_IAllocatorUniquePtr<T> MakeUniquePtr(std::shared_ptr<Provider_IAllocator> allocator, size_t count_or_bytes) {
+    if (allocator == nullptr) return nullptr;
+    // for now limit to fundamental types. we could support others, but to do so either we or the caller
+    // needs to call the dtor for the objects, for buffers allocated on device we don't have destructor
+    //static_assert(std::is_fundamental<T>::value, "Fundamental type required as no destructors are called.");
+
```

Comment:
Delete old comment that no longer applies

Reply:
Done